### PR TITLE
Add COCOTB_RESULTS_FILE environment variable

### PR DESCRIPTION
**Pull request by themperek**
Add `COCOTB_RESULTS_FILE_NAME` environment variable that can overwrite default "results.xml"

It became useful for more advanced tests and more pythonic unit test structure (also in the context of cocotb-test)

### CODE REVIEW

**Comment by cmarqu** [comment]
See also https://github.com/cocotb/cocotb/issues/508#issuecomment-514409648 by @jahagirdar

**Comment by jahagirdar** [comment]
I prefer COCOTB_RESULT_FILE_**PATH** over  COCOTB_RESULT_FILE_**NAME**

`filename=os.path.join(os.getenv("COCOTB_RESULT_FILE_PATH"),"results.xml")`

After running a regression I typically run `combine_results.py`which searches for "results.xml" through the given hierarchy. If we allow changing the xml filename, this script will need an update.

Note: Both options would break the `results.xml` target in simulator makefiles 

My personal version of this change is available at https://github.com/dyumnin/cocotb/blob/results_xml/cocotb/regression.py

**Comment by themperek** [comment]
@jahagirdar You can always do 

`COCOTB_RESULTS_FILE_NAME  = os.getenv("COCOTB_RESULT_FILE_PATH", "")+ "results.xml"` 

beacse of this `COCOTB_RESULTS_FILE_NAME` is more general.

**Comment by eric-wieser** [comment]
> Note: Both options would break the results.xml target in simulator makefiles

This is a really good call-out. Should this be a makefile-level variable, and used in the target name? Or should it be copied to the destination as an extra makefile step

**Comment by themperek** [comment]
It will only break if someone starts to set `COCOTB_RESULTS_FILE_NAME` but in this case should know what is doing and adjust results parsing/merging.

**Comment by themperek** [comment]
@imphil What I should change ? This does not break anything?

**Comment by imphil** [comment]
All makefiles currently use "results.xml" as target. If you change the name of this file, you also need to use the new file name as target name, or your Makefile dependencies are broken. Simply replace results.xml in the Makefile with your variable.

**Comment by jahagirdar** [comment]
Also an update to https://github.com/cocotb/cocotb/blob/master/documentation/source/building.rst may be required to document this variable.
Note: We may have to document all the environment variables found in this file.

**Comment by themperek** [comment]
Nothing changes it is "results.xml" by default. I would avoid changing all make files. This is special case that I would need to make parallel/iterative runs for coverage. Other way need to find some ugly workaround.

**Comment by themperek** [comment]
Nothing changes it is "results.xml" by default. I would avoid changing all make files. This is special case that I would need to make parallel/iterative runs for coverage. Other way need to find some ugly workaround.

**Comment by mciepluc** [comment]
what if results.xml remains named as it is, but we only change path to where it should be saved?
for parallel runs it will be required to create multiple results directories anyway

**Comment by imphil** [comment]
There are two tests I apply to a change:
1. Does it regress existing functionality?
2. Does the new functionality work, even in corner cases?

1 is clearly the case, as you pointed out: as long as one doesn't use COCOTB_RESULTS_FILE_NAME, everything stays as-is. Good.

2 is not the case. If one uses COCOTB_RESULTS_FILE_NAME with the standard makefiles, the dependency logic is broken. That's not really acceptable. The solution is rather simple: use $COCOTB_RESULTS_FILE_NAME in the Makefiles, instead of `results.xml`. 

Can you give that a try? It shouldn't be a difficult change to make.

**Comment by cmarqu** [comment]
I would also prefer ``COCOTB_RESULTS_FILE`` over ``COCOTB_RESULTS_FILE_NAME`` since it's clear here that we are not going to hand over a file handle or some such.

File: documentation/source/building.rst [hunk 1]
```diff
@@ -12,8 +12,7 @@ Make Targets
 
 Makefiles define two targets, ``regression`` and ``sim``, the default target is ``sim``.
 
-Both rules create a results file in the calling directory called :file:`results.xml`.  This file is a JUnit-compatible output file suitable for use with `Jenkins <https://jenkins.io/>`_. The ``sim`` targets unconditionally re-runs the simulator whereas the ``regression`` target only re-builds if any dependencies have changed.
-
+Both rules create a results file accordingly to :envvar:`COCOTB_RESULTS_FILE` variable.  This file is a JUnit-compatible output file suitable for use with `Jenkins <https://jenkins.io/>`_. The ``sim`` targets unconditionally re-runs the simulator whereas the ``regression`` target only re-builds if any dependencies have changed.
```
Comment (cmarqu) [inline documentation/source/building.rst]: I'm not a native speaker, but maybe 
```Both rules create a results file called :file:`results.xml` or, if set, whatever name :envvar:`COCOTB_RESULTS_FILE` specifies.```
is a bit nicer?

**Comment by imphil** [comment]
Thanks for this update @themperek! 

You need to update `tests/test_cases/issue_253/Makefile` to replace the results.xml target with your new target, and then this is ready to go!

**Comment by imphil** [comment]
Thanks, squashed and merged.

**Comment by themperek** [comment]
@imphil Thank you.
